### PR TITLE
[aten] Split some at::launch code into at::internal::launch_no_thread_state()

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38477 [aten] Split some at::launch code into at::internal::launch_no_thread_state()**

A few specific uses (e.g. Thrift rpc parsing) don't need source thread
state to be copied over. In microbenchmarks, this seems to add ~500ns,
so split code across functions, so some code can use directly.

Differential Revision: [D21573168](https://our.internmc.facebook.com/intern/diff/D21573168/)

### CODE REVIEW

**Comment by jjlilley** [comment]
fwiw: the motivation for this change is simply to improve efficiency of D21409536

File: aten/src/ATen/Parallel.h [hunk 1]
```diff
@@ -128,6 +128,9 @@ CAFFE2_API int get_num_interop_threads();
 
 // Launches inter-op parallel task
 CAFFE2_API void launch(std::function<void()> func);
+namespace internal {
+  void launch_no_thread_state(std::function<void()> fn);
```
Comment (wanchaol) [inline aten/src/ATen/Parallel.h:132]: nit: align the `func` name with `launch`

Comment (jjlilley) [reply aten/src/ATen/Parallel.h]: sorry, clash of different style guidelines in my head :) fixed.

**Review by wanchaol — APPROVED** [review]
looks good to me.

**Review by ilia-cher — APPROVED** [review]

**Pull request closed without merging**